### DISCUSSION TITLE
[BUG?] MediaElement Player Invalid State Called After Teardown

### DISCUSSION THREAD

**Discussion by dorton21** [CommunityToolkit/Maui#1900] [Q&A]
First, why isn't this under the bug section!? Simple, I cannot give the steps necessary to replicate this error due to the necessary hardware and privacy required for my project. I hope people will still be open to guiding me however as I could really use the advice.

**Context:** I have an RTSP stream over an Ubuntu Server my Android Tablet is connected to, while monitoring the wlan0 packets on the Ubuntu Server I can see the following commands get called OPTIONS, DESCRIBE, SETUP, SETUP, PLAY, TEARDOWN, SETUP... oops we've hit an error. The error message is "Error 455 Method Not Valid in this State". My assumption being SETUP cannot be called after teardown. Now you may be thinking to yourself, what if this isn't an issue with MediaElement and rather an issue with the RTSP stream on the Ubuntu Server (perhaps hostapd is acting up?) Well unfortunately after weeks of trial and error I have eliminated those as the culprits. The smoking gun being if I switch to the VLC app on the same tablet connected to the same Ubuntu Server consuming the same RTSP stream, the video plays back nicely.

**So What Do I Need:** 
If someone could point me in the right direction, I am more than happy to figure out the pull request that will fix this bugaboo (assuming it is in fact a bug with MediaElement). My assumption (based on limited RTSP knowledge) is before SETUP can be called again OPTIONS and DESCRIBE need to be called. I was wondering where I would tweak the code to do so since the closest thing I can find to accessing these calls it through the EventManager but even then it's not calling them directly. _OR_ perhaps I am going down the wrong rabbit hole and someone else has suggestions as to what the true bond villain may be.

I am open to all feedback and suggestions as I have been hunting this wabbit for weeks now. Thank you in advance!

**Comment by dorton21** [comment]
One thing I forgot to note above. I am making an underlying assumption that the bug is after a failed initial stream setup, there is an attempt to restart the setup. Instead of walking through the appropriate steps it jumps ahead.

In other words it goes: OPTIONS, DESCRIBE, SETUP, PLAY, TEARDOWN, SETUP
But it should go: OPTIONS, DESCRIBE, SETUP, PLAY, TEARDOWN, OPTIONS, DESCRIBE, SETUP, PLAY

I do not know this to be definitive as my RTSP familiarity is limited to Wikipedia articles and ChatGPT however. So if you think this is not the case please let me know.

**Comment by dorton21** [comment]
I have done further investigating and I did a comparison of the packets coming through on when I play with MediaElement vs playing with VLC. On the MediaElement there appears to be potential authentication issues? After a call to DESCRIBE in MediaElement I notice two lines that state 0 Refused 0/0/0 (33). Whereas this is not present under the DESCRIBE of the VLC. Both return 200 OK however. Further, I am curious as to why PLAY is called on VLC and the frames begin to come through just fine. Whereas PLAY is called on MediaElement and (despite returning 200 OK) I see a couple more 0 Refused 0/0/0 (33).

**Reply by bijington** [reply]
Have you tried enabling logging as per this comment in an old bug: https://github.com/CommunityToolkit/Maui/issues/972#issuecomment-1429635754

**Reply by dorton21** [reply]
Thanks for the tip. Upon reviewing this comment and its accompanying thread this does look to be very similar. Unfortunately, the steps Anioz took to solving this issue are not well documented. I can make a fair guess without enabling the debugger that I too am facing authentication issues. This would imply I need to tweak something relative to the exoplayer code but I am not certain how to even find the code locally since I am dealing with the nuget packaged version of media player.

**Reply by bijington** [reply]
An option there would be to fork/clone the toolkit and either:
1) add your server usage into the sample application and debug with the toolkit code
2) add the toolkit projects into your project and use that way rather than as a nuget package

Both will allow you to debug into the toolkit code
